src/material/tree/node.ts

```diff
@@ -25,36 +25,106 @@ import {
   providers: [{provide: CdkTreeNode, useExisting: MatTreeNode}],
   host: {
     'class': 'mat-tree-node',
+    '[attr.aria-expanded]': '_getAriaExpanded()',
+    '[attr.aria-level]': 'level + 1',
+    '[attr.aria-posinset]': '_getPositionInSet()',
+    '[attr.aria-setsize]': '_getSetSize()',
+    '(click)': '_focusItem()',
+    'tabindex': '_getTabindexAttribute()',
   },
 })
 export class MatTreeNode<T, K = T> extends CdkTreeNode<T, K> implements OnInit, OnDestroy {
-  /** Whether the node is disabled. */
-  @Input({transform: booleanAttribute})
-  disabled: boolean = false;
+  /**
+   * The tabindex of the tree node.
+   *
+   * @deprecated By default MatTreeNode manages focus using TreeKeyManager instead of tabIndex.
+   *   Recommend to avoid setting tabIndex directly to prevent TreeKeyManager form getting into
+   *   an unexpected state. Tabindex to be removed in a future version.
+   * @breaking-change 19.0.0 Remove this attribute.
+   */
 
-  /** Tabindex of the node. */
   @Input({
     transform: (value: unknown) => (value == null ? 0 : numberAttribute(value)),
   })
-  tabIndex: number;
+  get tabIndex(): number {
+    return this.isDisabled ? -1 : this._tabIndex;
+  }
+  set tabIndex(value: number) {
+    // If the specified tabIndex value is null or undefined, fall back to the default value.
+    this._tabIndex = value;
+  }
+  private _tabIndex: number;
+
+  /**
+   * The default tabindex of the tree node.
+   *
+   * @deprecated By default MatTreeNode manages focus using TreeKeyManager instead of tabIndex.
+   *   Recommend to avoid setting tabIndex directly to prevent TreeKeyManager form getting into
+   *   an unexpected state. Tabindex to be removed in a future version.
+   * @breaking-change 19.0.0 Remove this attribute.
+   */
+  defaultTabIndex = 0;
+
+  protected _getTabindexAttribute() {
+    if (isLegacyTreeKeyManager(this._tree._keyManager)) {
+      return this.tabIndex;
+    }
+    return -1;
+  }
+
+  /**
+   * Whether the component is disabled.
+   *
+   * @deprecated This is an alias for `isDisabled`.
+   * @breaking-change 19.0.0 Remove this input
+   */
+  @Input({transform: booleanAttribute})
+  get disabled(): boolean {
+    return this.isDisabled;
+  }
+  set disabled(value: boolean) {
+    this.isDisabled = value;
+  }
 
   constructor(
     elementRef: ElementRef<HTMLElement>,
     tree: CdkTree<T, K>,
+    /**
+     * The tabindex of the tree node.
+     *
+     * @deprecated By default MatTreeNode manages focus using TreeKeyManager instead of tabIndex.
+     *   Recommend to avoid setting tabIndex directly to prevent TreeKeyManager form getting into
+     *   an unexpected state. Tabindex to be removed in a future version.
+     * @breaking-change 19.0.0 Remove this attribute.
+     */
     @Attribute('tabindex') tabIndex: string,
   ) {
     super(elementRef, tree);
-    this.tabIndex = Number(tabIndex) || 0;
+
+    this.tabIndex = Number(tabIndex) || this.defaultTabIndex;
```

Comment:
Is this correct? This will assign `defaultTabIndex` for `tabindex="0"`. If we want to respect the users's explicit 0 setting, we could do: `this.tabIndex = tabIndex ? Number(tabIndex) : this.defaultTabIndex`

Reply:
Intention is that the implementation is exactly the same as before, but it's refactored to not use the `hasTabindex` mixin.